Rename browser to driver

aaltat opened the pull request:
To do:
- [x] fix browsermanagement.py, because it says `browser` in many places where it should say `driver`
- [x] deprecate method `self.browser`

aaltat committed:
rename browsercache.py to drivercache.py

aaltat committed:
Fixed DriverCache internal from browser to driver

aaltat committed:
Changed in code browser(s) to driver(s)

aaltat committed:
Fixing code to pass tests

aaltat commented:
But it definitely looks like my rebase with master did not go smoothly. Have to take a look later.

pekkaklarck commented:
I'd still use  `webdriver` instead of `driver` as the attribute name, but it seems everyone else likes the latter so let's use that. I'd still use use `WebDriverCache` and not `DriverCache`, though. More importantly, the type of the attribute (i.e. property) should be specified in the docstring. That makes the attribute type unambiguous for users and also helps IDEs.

aaltat committed:
Renamed DriverCache to WebDriverCache

aaltat committed:
Added documentation for driver public API.

Also created public API browser for RC1 to support early adopters.

aaltat committed:
Refactoring BrowserManagementKeywords browser -> driver

Contains changes in the internal variables and method names.

aaltat commented:
The `windowmanager.py` needs some refactoring, so that it does not pass the `driver` as argument but uses the `self.driver` instead

pekkaklarck commented on src/SeleniumLibrary/__init__.py:
```diff
@@ -343,7 +343,7 @@ def __init__(self, timeout=5.0, implicit_wait=0.0,
             WindowKeywords(self)
         ]
-        self._browsers = BrowserCache()
+        self._webdrivers = WebDriverCache()
```
This should be `_drivers` to match ` driver`.

aaltat replied:
Done

pekkaklarck commented on src/SeleniumLibrary/base/context.py:
```diff
@@ -28,12 +28,12 @@ def __init__(self, ctx):
     @property
-    def browsers(self):
-        return self.ctx._browsers
+    def webdrivers(self):
```
This should be `drivers` to match ` driver`.

aaltat replied:
Done

pekkaklarck commented on src/SeleniumLibrary/keywords/browsermanagement.py:
```diff
@@ -368,8 +367,8 @@ def set_selenium_speed(self, value):
         self.ctx.speed = timestr_to_secs(value)
-        for browser in self.browsers.browsers:
-            self._monkey_patch_speed(browser)
+        for driver in self.webdrivers.drivers:
```
Isn't ConnectionCache iterable? If it is, this could be just `for driver in self.drivers`. If it isn't, it should be changed!

aaltat replied:
It should be iterable, but changes the code to only iterate drivers which are not closed.

aaltat committed:
Refactor WindowManager to use self.driver

The WindowManager used to pass driver between methods. Now the
methods use self.driver, instead of passing the diver as argument.

aaltat committed:
Updated based on review comments

pekkaklarck commented on src/SeleniumLibrary/__init__.py:
```diff
@@ -385,11 +385,22 @@ def failure_occurred(self):
 
     @property
+    def driver(self):
+        """Current active driver
```
`.` missing at the end.

aaltat replied:
Done

pekkaklarck commented on src/SeleniumLibrary/__init__.py:
```diff
@@ -385,11 +385,22 @@ def failure_occurred(self):
+        :raises RuntimeError if driver is not created
+        """
+        if not self._drivers.current:
+            raise RuntimeError('No driver is open')
```
Strange error message. Perhaps this should still be `No browser is open.`? Also, should we use a dedicated exception?

aaltat replied:
Yes, browser is better.

It's strange exception, but that is how the original code worked. Dedicated exception would be better, would `NoBrowser` be good?

pekkaklarck replied:
Perhaps NoOpenBrowser or NoBrowserOpen?

pekkaklarck commented on src/SeleniumLibrary/utils/webdrivercache.py:
```diff
@@ -17,33 +17,33 @@
-            if browser not in self._closed:
-                open_browsers.append(browser)
-        return open_browsers
+    def get_open_drivers(self):
```
Open driver sounds a bit odd. Should this perhaps be `get_active_browsers()`? Or perhaps property `active_browsers`?

aaltat replied:
Did go with `active_drivers`

pekkaklarck reviewed (approved):
Looks great. Added few line notes, though.

aaltat committed:
Small adjustments based on comments.

pekkaklarck commented on src/SeleniumLibrary/utils/webdrivercache.py:
```diff
@@ -17,33 +17,34 @@
 
     def __init__(self):
-        ConnectionCache.__init__(self, no_current_msg='No current browser')
+        ConnectionCache.__init__(self, no_current_msg='No current driver')
```
This should still say `browser`. And have `.` at the end.

If that message is ever used, it's used by `ConnectionCache` as a message for `RuntimeError`. We have our own `NoOpenBrowser` now, but there's no good API to use that here. We could set `self._no_current` to an object using our exception (and it could extend `NoConnection`) but I'm not sure is that worth the effort.

aaltat replied:
I did change to driver -> browser. 

If the start seeing the `RuntimeError`in real life, lets then decide what we should do.

pekkaklarck commented on src/SeleniumLibrary/__init__.py:
```diff
@@ -384,11 +385,22 @@ def failure_occurred(self):
+        :raises RuntimeError if driver is not created
+        """
+        if not self._drivers.current:
+            raise NoOpenBrowser('No browser is open')
```
- Exception type not updated in the doc string above.
- This error message could have `.` at the end.

aaltat replied:
Done.

aaltat committed:
Fixed doc string and error message in cache

aaltat committed:
Fixed tests because error message was changed

aaltat committed:
Fixes sphinx doc definition

pekkaklarck commented on src/SeleniumLibrary/__init__.py:
```diff
@@ -384,11 +385,22 @@ def failure_occurred(self):
+        """Current active driver.
+
+        :rtype: selenium.webdriver.remote.webdriver.WebDriver
+        :raises SeleniumLibrary.errors.ElementNotFound: If element not found.
```
Thought it would raise NoOpenBrowser....

aaltat committed:
Fixes agains the doc string
